### PR TITLE
refactor: rewrite VaForm.vue with context api

### PR DESCRIPTION
**Pull request by LighthouseKeeperYN**
close: #1062

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

### CODE REVIEW

File: packages/ui/src/components/va-form/consts.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
 import { InjectionKey } from 'vue'
 
 export type FormChild = {
-  hasError?: () => any;
-  reset: () => any;
-  resetValidation: () => any;
-  focus: () => any;
+  hasError?: () => boolean;
```
Comment (asvae) [inline packages/ui/src/components/va-form/consts.ts:4]: Curious why this is optional?

File: packages/ui/src/components/va-form/consts.ts [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
+  reset: () => void;
+  resetValidation: () => void;
+  focus: () => void;
   validate: () => boolean;
+  focusInvalid?: () => void;
```
Comment (asvae) [inline packages/ui/src/components/va-form/consts.ts:9]: And this. :)

**Review by asvae — APPROVED** [review]

File: packages/ui/src/components/va-form/VaForm.vue [hunk 3]
```diff
@@ -8,115 +9,120 @@
...
 
-class FormProps {
-  autofocus = prop<boolean>({ type: Boolean, default: false })
-  tag = prop<string>({ type: String, default: 'div' })
-}
+  setup (props, { emit }) {
+    const nestedFormElements: Ref<FormChild[]> = ref([])
 
-const FormPropsMixin = Vue.with(FormProps)
+    const parentFormProvider = () => ({ ...inject(FormServiceKey, undefined) })
```
Comment (RVitaly1978) [inline packages/ui/src/components/va-form/VaForm.vue:39]: Can we do this?:
`const parentFormProvider = inject(FormServiceKey, undefined)`

File: packages/ui/src/components/va-form/VaForm.vue [hunk 4]
```diff
@@ -8,115 +9,120 @@
...
+    }
 
-  childUnmountedHandler (removableChild: FormChild | VaForm) {
-    this.nestedFormElements = this.nestedFormElements.filter(child => child !== removableChild)
-  }
+    const focus = () => {
+      const focusableElement = nestedFormElements.value.find(({ focus }) => focus)
+      if (focusableElement) {
+        focusableElement.focus()
+      }
```
Comment (RVitaly1978) [inline packages/ui/src/components/va-form/VaForm.vue:71]: Maybe we can:
`nestedFormElements.value.find(({ focus }) => focus)?.focus()`

File: packages/ui/src/components/va-form/VaForm.vue [hunk 5]
```diff
@@ -8,115 +9,120 @@
...
-  focus () {
-    const focusableElement = this.nestedFormElements.find(({ focus }) => focus)
-    if (focusableElement) {
-      focusableElement.focus()
+      if (invalidComponent) {
+        invalidComponent.focus()
+      } else {
+        nestedFormElements.value
+          .forEach(item => isVaForm(item) && item.focusInvalid?.())
+      }
```
Comment (RVitaly1978) [inline packages/ui/src/components/va-form/VaForm.vue:83]: If we have
`const isVaForm = (value: any) => !!value.focusInvalid`
So we don't need to check for the existence of the method:
`item.focusInvalid?.()` ==> `item.focusInvalid()`

**Review by RVitaly1978 — APPROVED** [review]
Things are good. Just take a look at some of my suggestions.

**Pull request merged**